src/payment/strategies/masterpass/masterpass-payment-strategy.ts

```diff
@@ -0,0 +1,121 @@
+import { PaymentStrategy } from '../';
+import {
+    PaymentActionCreator,
+    PaymentInitializeOptions,
+    PaymentMethodActionCreator,
+    PaymentRequestOptions
+} from '../../';
+import { Checkout, CheckoutStore, InternalCheckoutSelectors } from '../../../checkout';
+import {
+    InvalidArgumentError,
+    MissingDataError,
+    MissingDataErrorType
+} from '../../../common/error/errors';
+import { StoreConfig } from '../../../config';
+import { OrderActionCreator, OrderRequestBody } from '../../../order';
+import PaymentMethod from '../../payment-method';
+
+import { MasterpassCheckoutCallback, MasterpassCheckoutOptions } from './masterpass';
+import MasterpassScriptLoader from './masterpass-script-loader';
+
+export default class MasterpassPaymentStrategy extends PaymentStrategy {
+    private _paymentMethod?: PaymentMethod;
+
+    constructor(
+        store: CheckoutStore,
+        private _orderActionCreator: OrderActionCreator,
+        private _paymentActionCreator: PaymentActionCreator,
+        private _paymentMethodActionCreator: PaymentMethodActionCreator,
+        private _masterpassScriptLoader: MasterpassScriptLoader
+    ) {
+        super(store);
+    }
+
+    initialize(options: PaymentInitializeOptions): Promise<InternalCheckoutSelectors> {
+        const { methodId } = options;
+
+        this._paymentMethod = this._store.getState().paymentMethods.getPaymentMethod(methodId);
+
+        if (!this._paymentMethod) {
+            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+        }
+
+        if (!this._hasPaymentInfo()) {
+            return this._masterpassClientSetup().then(checkoutCallback => checkoutCallback()).then(() => super.initialize(options));
```

Comment:
Thank you for the screen capture. That video shows exactly what it shouldn't happen. We shouldn't be redirecting the shopper in this method. As I said before, this method is only responsible for initialising the payment method (i.e.: loading the Masterpass client JS); not prompting the shopper to enter their payment details.

Reply:
@davidchin Basically if we move the logic to the execute the flow will change

This is the expected flow

1. The user goes to the checkout page
2. It selects masterpass as payment method
3. It gets redirected to the masterpass page to pick a card
4. It returns to bcapp checkout page, the wallet is displayed.
5. The user click on place order (payment strategy execute method) and the order and the payment is created.

If we move the redirect logic to the execute method:

1. The user goes to the checkout page
2. It selects masterpass as payment option
3. The user click on place order (payment strategy execute method)
4. It gets redirected to the masterpass page to pick a card.
5. It returns to bcapp checkout page, the wallet is displayed.
6. The user click on place order again (payment strategy execute method) and the order and the payment is created.

FYI: @Sure-Round